libr/util/logging.c

```diff
@@ -0,0 +1,123 @@
+static char cfg_logfile[LOGGING_CONFIGSTR_SIZE] = ""; // Output text to filename
+static const char *level_tags[] = { // Log level to tag string lookup array
+	[R_LOGLVL_SILLY]     = "SILLY: ",
+	[R_LOGLVL_VERBOSE]   = "VERBOSE: ",
+	[R_LOGLVL_DEBUG]     = "DEBUG: ",
+	[R_LOGLVL_INFO]      = "INFO: ",
+	[R_LOGLVL_WARN]      = "WARNING: ",
+	[R_LOGLVL_ERROR]     = "ERROR: ",
+	[R_LOGLVL_FATAL]     = "FATAL: ",
+	[R_LOGLVL_FATALTRAP] = "FATAL: "
+};
+
+// cconfig.c configuration callback functions below
+R_API void r_logging_set_level(RLoggingLevel level) {
+	cfg_loglvl = level;
+}
+
+R_API void r_logging_set_traplevel(RLoggingLevel level) {
+	cfg_logtraplvl = level;
+}
+
+R_API void r_logging_set_file(const char *filename) {
+	int value_len = strnlen (filename, LOGGING_CONFIGSTR_SIZE) + 1;
+	strncpy (cfg_logfile, filename, value_len);
+}
+
+R_API void r_logging_set_srcinfo(bool show_info) {
+	cfg_logsrcinfo = show_info;
+}
+
+R_API void r_logging_set_colors(bool show_info) {
+	cfg_logcolors = show_info;
+}
+
+/**
+ * \brief Set the logging callback function for printing
+ * \param cbfunc RLoggingFuncdef style function to be called
+*/
+R_API void r_logging_set_callback(RLoggingFuncdef cbfunc) {
+	// TODO: RList of callbacks with setter/remove methods
+	cb_logfunc_print = cbfunc;
+}
+
+/**
+ * \brief Internal logging function used by preprocessor macros
+ * \param funcname Contains the function name of the calling function
+ * \param filename Contains the filename that funcname is defined in
+ * \param lineno The line number that this log call is being made from in filename
+ * \param lvl Logging level for output
+ * \param fmtstr A printf like string
+ * This function is used by the R_LOG_* preprocessor macros for logging/debugging
+*/
+R_API void _r_logging_internal(const char *funcname, const char *filename,
+	ut32 lineno, ut32 level, const char *fmtstr, ...) {
+	if (level < cfg_loglvl && level < cfg_logtraplvl) {
+		// Don't print if output level is lower than current level
+		// Don't ignore fatal/trap errors
+		return;
+	}
+
+	// Setup varadic arguments
+	va_list args;
+	va_start (args, fmtstr);
+
+	// TODO: Colors
+
+	// Build output string with src info, and formatted output
+	char output_buf[LOGGING_OUTPUTBUF_SIZE] = ""; // Big buffer for building the output string
+	const char *tag = R_BETWEEN (0, level, sizeof (level_tags) - 1) ? level_tags[level] : "";
+	int offset = snprintf (output_buf, LOGGING_OUTPUTBUF_SIZE, "%s", tag);
+	if (cfg_logsrcinfo) {
+		offset += snprintf (output_buf + offset, LOGGING_OUTPUTBUF_SIZE - offset, "%s in %s:%i: ", funcname, filename, lineno);
+	}
+	vsnprintf (output_buf + offset, LOGGING_OUTPUTBUF_SIZE - offset, fmtstr, args);
+
+	// Actually print out the string with our callbacks
+	if (cb_logfunc_print) {
+		cb_logfunc_print (output_buf, funcname, filename, lineno, level, fmtstr, args);
+	} else {
+		if (level < R_LOGLVL_ERROR) {
+			fprintf (stdout, "%s", output_buf);
```

Comment:
imho. all log messages should go to stderr

Reply:
but. maybe we should have a callback to handle how to. print log messages (from core where this thing will be instantiated) and allow to send via udp sockets, use r_log, or whatever (cutter?) decides

Reply:
This is the fallback. There is already the callback to choose to print it however you want, but it still makes sense to have a fallback solution when there's no callback setup (for example if you are using this before the callback is set or in other projects where r_core is not used)